ls1intum/Artemis
# Code-Editor/Conflict resolution

### Checklist

### Motivation and Context



When working both from a local git repository and in the Artemis online code editor, CheckoutConflicts can be created that make the code editor unusable.
This feature is supposed to provide a tool that helps students to resolve such conflicts.

2 scenarios, depending on the severity of the conflict: 
- a) Creates a merge conflict. When reloading the page, the user will see conflict markers in the files. The issue is here that when saving files from the editor BEFORE reloading the page, it will overwrite the changes from the remote.
- b) Makes the editor unusable as all pull actions fail. In this scenario the user would see a broken editor.

### Description


This PR adds handling of conflict exceptions on the server, and offering a tool in the client to resolve conflicts.

The point where the exceptions occurr is `GitService.pull`, so Exceptions will bubble up from the `GitService` to the `RepositoryResource`. 
This needs to be done because all actions (save, create, commit, etc.) could cause a checkout exception. (*see more info on this below)
`RepositoryResource.getStatus` does additional checking when the code editor is opened to make sure that no unnecessary calls are executed (e.g. get files when a conflict exists).

*I thought about changing the online editor to only pull on a commit. However this would mean that the user could work a long time in the online editor, then commit with the result of a conflict, reload the page and not being able to retrieve the staged changes as the `getFiles` call will fail with a `WrongRepositoryState` exception. So we should try to catch the conflict as early as possible so the user does not lose to much of their work.

- [x] server implementation
- [x] client implementation
- [x] update existing jest tests
- [x] write server tests
- [x] write new jest component tests
- [x] write new jest integration tests
- [x] server refactoring
- [x] cleanup & documentation

### Steps for Testing


How to reproduce a checkout conflict exception:

- Pull git repo locally
- Change line in code, stage, commit
- Change same line in artemis editor, save
- Push local commit
- Submit in the Artemis editor
=> CheckoutConflictException
https://stackoverflow.com/questions/18050220/eclipse-egit-checkout-conflict-with-files-egit-doesnt-want-to-continue

After that resolve the conflict and see if everything works again :)

### Screenshots



![image](https://user-images.githubusercontent.com/28230611/60773381-08705a00-a105-11e9-84a4-070dd84ab403.png)

![image](https://user-images.githubusercontent.com/28230611/60773384-2211a180-a105-11e9-8b26-ae654c76c146.png)
